pkg/registry/client.go

```diff
@@ -0,0 +1,161 @@
+	}
+)
+
+// NewClient returns a new registry client with config
+func NewClient(options *ClientOptions) *Client {
+	return &Client{
+		out:      options.Out,
+		resolver: options.Resolver,
+		cache: &filesystemCache{
+			out:     options.Out,
+			rootDir: options.CacheRootDir,
+			store:   orascontent.NewMemoryStore(),
+		},
+	}
+}
+
+// PushChart uploads a chart to a registry
+func (c *Client) PushChart(ref *Reference) error {
+	c.setDefaultTag(ref)
+	fmt.Fprintf(c.out, "The push refers to repository [%s]\n", ref.Locator)
+	layers, err := c.cache.LoadReference(ref)
+	if err != nil {
+		return err
+	}
+	err = oras.Push(context.Background(), c.resolver, ref.String(), c.cache.store, layers)
+	if err != nil {
+		return err
+	}
+	var totalSize int64
+	for _, layer := range layers {
+		totalSize += layer.Size
+	}
+	fmt.Fprintf(c.out,
+		"%s: pushed to remote (%d layers, %s total)\n", ref.Object, len(layers), byteCountBinary(totalSize))
+	return nil
+}
+
+// PullChart downloads a chart from a registry
+func (c *Client) PullChart(ref *Reference) error {
+	c.setDefaultTag(ref)
+	fmt.Fprintf(c.out, "%s: Pulling from %s\n", ref.Object, ref.Locator)
+	layers, err := oras.Pull(context.Background(), c.resolver, ref.String(), c.cache.store, KnownMediaTypes()...)
+	if err != nil {
+		return err
+	}
+	exists, err := c.cache.StoreReference(ref, layers)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		fmt.Fprintf(c.out, "Status: Downloaded newer chart for %s:%s\n", ref.Locator, ref.Object)
+	} else {
+		fmt.Fprintf(c.out, "Status: Chart is up to date for %s:%s\n", ref.Locator, ref.Object)
+	}
+	return nil
+}
+
+// SaveChart stores a copy of chart in local cache
+func (c *Client) SaveChart(ch *chart.Chart, ref *Reference) error {
+	c.setDefaultTag(ref)
+	layers, err := c.cache.ChartToLayers(ch)
+	if err != nil {
+		return err
+	}
+	_, err = c.cache.StoreReference(ref, layers)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(c.out, "%s: saved\n", ref.Object)
+	return nil
+}
+
+// LoadChart retrieves a chart object by reference
+func (c *Client) LoadChart(ref *Reference) (*chart.Chart, error) {
+	c.setDefaultTag(ref)
+	layers, err := c.cache.LoadReference(ref)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := c.cache.LayersToChart(layers)
+	return ch, err
+}
+
+// RemoveChart deletes a locally saved chart
+func (c *Client) RemoveChart(ref *Reference) error {
+	c.setDefaultTag(ref)
+	err := c.cache.DeleteReference(ref)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(c.out, "%s: removed\n", ref.Object)
+	return err
+}
+
+// PrintChartTable prints a list of locally stored charts
+func (c *Client) PrintChartTable() error {
+	table := uitable.New()
+	table.MaxColWidth = 60
+	table.AddRow("REF", "NAME", "VERSION", "DIGEST", "SIZE", "CREATED")
+	rows, err := c.cache.TableRows()
+	if err != nil {
+		return err
+	}
+	for _, row := range rows {
+		if len(row) == 6 {
+			table.AddRow(row[0], row[1], row[2], row[3], row[4], row[5])
```

Comment:
I believe it's possible to do something like `table.AddRow(row...)`, which will iterate through each item in the list and add it as an argument.

here's an example in practice: https://play.golang.org/p/1Dq-DH6GZhc

Reply:
@adamreese told me this isn't possible as `table.AddRow` accepts an `[]interface{}`. I had assumed wrongly that it accepted a string slice. My bad

Reply:
made it happen :)